secure_tunneling/include/aws/iotsecuretunneling/SecureTunnel.h

```diff
@@ -23,31 +25,76 @@ namespace Aws
         using OnStreamReset = std::function<void(void)>;
         using OnSessionReset = std::function<void(void)>;
 
-        class AWS_IOTSECURETUNNELING_API SecureTunnel final
+        class AWS_IOTSECURETUNNELING_API SecureTunnelBuilder final
         {
           public:
-            SecureTunnel(
-                // Parameters map to aws_secure_tunneling_connection_config
+            /**
+             * Constructor arguments are the minimum required to create a secure tunnel
+             */
+            SecureTunnelBuilder(
                 Crt::Allocator *allocator,                        // Should out live this object
-                Aws::Crt::Io::ClientBootstrap *clientBootstrap,   // Should out live this object
+                Aws::Crt::Io::ClientBootstrap &clientBootstrap,   // Should out live this object
                 const Aws::Crt::Io::SocketOptions &socketOptions, // Make a copy and save in this object
-                Aws::Crt::Http::HttpClientConnectionProxyOptions
-                    *httpClientConnectionProxyOptions, // Make a copy and save in this object
-
-                const std::string &accessToken, // Make a copy and save in this object
+                const std::string &accessToken,                   // Make a copy and save in this object
                 aws_secure_tunneling_local_proxy_mode localProxyMode,
-                const std::string &endpointHost, // Make a copy and save in this object
-                const std::string &rootCa,       // Make a copy and save in this object
+                const std::string &endpointHost); // Make a copy and save in this object
+
+            /**
+             * Optional members
+             */
+            SecureTunnelBuilder &WithRootCa(const std::string &rootCa);
+            SecureTunnelBuilder &WithHttpClientConnectionProxyOptions(
+                const Aws::Crt::Http::HttpClientConnectionProxyOptions &httpClientConnectionProxyOptions);
+            SecureTunnelBuilder &WithOnConnectionComplete(OnConnectionComplete onConnectionComplete);
+            SecureTunnelBuilder &WithOnConnectionShutdown(OnConnectionShutdown onConnectionShutdown);
+            SecureTunnelBuilder &WithOnSendDataComplete(OnSendDataComplete onSendDataComplete);
+            SecureTunnelBuilder &WithOnDataReceive(OnDataReceive onDataReceive);
+            SecureTunnelBuilder &WithOnStreamStart(OnStreamStart onStreamStart);
+            SecureTunnelBuilder &WithOnStreamReset(OnStreamReset onStreamReset);
+            SecureTunnelBuilder &WithOnSessionReset(OnSessionReset onSessionReset);
+
+            /**
+             * Will return a shared pointer to a new SecureTunnel that countains a
+             * new aws_secure_tunnel that is generated using the set members of SecureTunnelBuilder.
+             * Returns a nullptr on fail.
+             */
+            std::shared_ptr<SecureTunnel> Build() noexcept;
```

Comment:
add comment how to learn about what went wrong

On failure, the shared_ptr will contain a nullptr, call aws_last_error() to get the reason for failure.